contracts/src/PriceFeeds/CompositePriceFeed.sol

```diff
@@ -4,73 +4,72 @@ pragma solidity 0.8.18;
+
+        // If branch is already shut down and using ETH-USD * canonical_rate, try to use that
+        if (priceSource == PriceSource.ETHUSDxCanonical) {
+                (uint256 ethUsdPrice, bool ethUsdOracleDown) = _getOracleAnswer(ethUsdOracle);
+                //... but if the ETH-USD oracle *also* fails here, use the lastGoodPrice
+                if(ethUsdOracleDown) {
+                    // No need to shut down, since branch already is shut down
+                    return (lastGoodPrice, false);
+                } else {
+                    return (_fetchPriceETHUSDxCanonical(ethUsdPrice), false);
+                }
+            }
+
+        // Otherwise if branch is shut down and already using the lastGoodPrice, continue with it
+        if (priceSource == PriceSource.lastGoodPrice) {return (lastGoodPrice, false);}
     }
 
-    function _fetchPrice() internal override returns (uint256, bool) {
-        (uint256 ethUsdPrice, bool ethUsdOracleDown) = _getOracleAnswer(ethUsdOracle);
-        (uint256 lstEthPrice, bool lstEthOracleDown) = _getOracleAnswer(lstEthOracle);
-
-        // If one of Chainlink's responses was invalid in this transaction, disable this PriceFeed and
-        // return the last good LST-USD price calculated
-        if (ethUsdOracleDown) {return (_disableFeedAndShutDown(address(ethUsdOracle.aggregator)), true);}
-        if (lstEthOracleDown) {return (_disableFeedAndShutDown(address(lstEthOracle.aggregator)), true);}
+    function _shutDownAndSwitchToETHUSDxCanonical(address _failedOracleAddr, uint256 _ethUsdPrice) internal returns (uint256) {
+        // Shut down the branch
+        borrowerOperations.shutdownFromOracleFailure(_failedOracleAddr);
 
-        // Calculate the market LST-USD price: USD_per_LST = USD_per_ETH * ETH_per_LST
-        uint256 lstUsdMarketPrice = ethUsdPrice * lstEthPrice / 1e18;
+        priceSource = PriceSource.ETHUSDxCanonical;
+        return _fetchPriceETHUSDxCanonical(_ethUsdPrice);
+    }
 
+    // Only called if the primary LST oracle has failed, branch has shut down, 
+    // and we've switched to using: ETH-USD * canonical_rate.
+    function _fetchPriceETHUSDxCanonical(uint256 _ethUsdPrice) internal returns (uint256) {
+        assert(priceSource == PriceSource.ETHUSDxCanonical);
         // Get the ETH_per_LST canonical rate directly from the LST contract
         // TODO: Should we also shutdown if the call to the canonical rate reverts, or returns 0?
         uint256 lstEthRate = _getCanonicalRate();
 
         // Calculate the canonical LST-USD price: USD_per_LST = USD_per_ETH * ETH_per_LST
-        uint256 lstUsdCanonicalPrice = ethUsdPrice * lstEthRate / 1e18;
-
-        // Take the minimum of (market, canonical) in order to mitigate against upward market price manipulation
-        uint256 lstUsdPrice = LiquityMath._min(lstUsdMarketPrice, lstUsdCanonicalPrice);
+        uint256 lstUsdCanonicalPrice = _ethUsdPrice * lstEthRate / 1e18;
```

Comment:
I wonder if it would still make sense to use the min against `lastGoodPrice`.
If the branch was shutdown, it may be due to a bug in the LST, so the canonical price may be wrong.
This way we would at least make sure we protect against the worst case of urgent redemptions not being profitable.

Reply:
Right, yes this makes sense. It could still be that `market_price < min(lastGoodPrice, canonical_price)` (depends on price movement) - but it increases the chances that we can fully clear the bad debt. Good idea!

Reply:
This is not done yet, right?